src/gt4py/next/common.py

```diff
@@ -61,21 +65,167 @@ def __str__(self):
         return f'Dimension(value="{self.value}", kind={self.kind})'
 
 
-class DType:
-    ...
+DomainLike = Union[Sequence[Dimension], Dimension, str]
 
 
-class Field(Generic[DimsT, DT]):
-    ...
+@dataclasses.dataclass(frozen=True)
+class UnitRange(Sequence, Set):
+    """Range from `start` to `stop` with step size one."""
 
+    start: int
+    stop: int
 
-@dataclass(frozen=True)
+    def __len__(self) -> int:
+        return max(0, self.stop - self.start)
+
+    def __repr__(self) -> str:
+        return f"UnitRange({self.start}, {self.stop})"
+
+    @overload
+    def __getitem__(self, index: int) -> int:
+        ...
+
+    @overload
+    def __getitem__(self, index: slice) -> list[int]:
+        ...
+
+    def __getitem__(self, index: int | slice) -> int | list[int]:
+        if isinstance(index, slice):
+            start, stop, step = index.indices(len(self))
+            return [self[i] for i in range(start, stop, step)]
+
+        if index < 0:
+            index += len(self)
+
+        if 0 <= index < len(self):
+            return self.start + index
+        else:
+            raise IndexError("UnitRange index out of range")
+
+
+DomainT: TypeAlias = tuple[tuple[Dimension, UnitRange], ...]
+
+
+class Field(Protocol[DimsT, gt4py_defs.ScalarT]):
+    @property
+    def domain(self) -> DomainT:
+        ...
+
+    @property
+    def dtype(self) -> DType[gt4py_defs.ScalarT]:
+        ...
+
+    @property
+    def value_type(self) -> gt4py_defs.ScalarT:
+        ...
+
+    @property
+    def ndarray(self) -> NDArrayObject:
+        ...
+
+    def __setattr__(self, key, value) -> None:
+        raise TypeError("Immutable type")
+
+    def __setitem__(self, key, value) -> None:
+        raise TypeError("Immutable type")
+
+    def __str__(self) -> str:
+        assert hasattr(self.value_type, "__name__")
+        codomain = (
+            f"{self.value_type!s}"
+            if isinstance(self.value_type, Dimension)
+            else self.value_type.__name__
+        )
+        return f"⟨{self.domain!s} → {codomain}⟩"
+
+    @abc.abstractmethod
+    def remap(self, index_field: Field) -> Field:
+        ...
+
+    @abc.abstractmethod
+    def restrict(self, item: "DomainLike") -> Field:
+        ...
+
+    # Operators
+    @abc.abstractmethod
+    def __call__(self, index_field: Field) -> Field:
+        ...
+
+    @abc.abstractmethod
+    def __getitem__(self, item: "DomainLike") -> Field:
+        ...
+
+    @abc.abstractmethod
+    def __abs__(self) -> Field:
+        ...
+
+    @abc.abstractmethod
+    def __neg__(self) -> Field:
```

Comment:
```suggestion
    def __neg__(self) -> Field[DimsT, ScalarT]:
```
we could be more specific for most of the operations, right?
just `remap` cannot be properly typed.